tools/kagami/src/genesis.rs

```diff
@@ -241,10 +259,231 @@ fn generate_synthetic(
+                })
+            }
+            KeyStorage::FromCLI(multihash_string) => PublicKey::from_str(multihash_string)
+                .wrap_err_with(|| {
+                    eyre!(
+                        "Failed to deserialize public key from multihash string for algorithm `{}`",
+                        self.algorithm
+                    )
+                }),
+        }
+    }
+}
+
+impl<T: Write> RunArgs<T> for SignArgs {
+    fn run(self, writer: &mut BufWriter<T>) -> Outcome {
+        let key_pair = if let Some(path) = self.keypair_file {
+            read_keypair(path)?
+        } else {
+            let (public_key, private_key) = (self.get_public_key()?, self.get_private_key()?);
+            KeyPair::new(public_key, private_key)?
+        };
+
+        let genesis_block = RawGenesisBlock::from_path(&self.genesis_file)?;
+        let genesis_signature =
+            GenesisNetwork::new_genesis_signature(genesis_block, &self.chain_id, &key_pair);
+        let genesis_signature_string = genesis_signature.to_hex_string();
+
+        let scale_encoded_signature = self.binary.then(|| genesis_signature_string.encode());
+
+        if let Some(path) = self.out_file {
+            if let Some(encoded_genesis_signature) = scale_encoded_signature {
+                fs::write(&path, encoded_genesis_signature)?;
+            } else {
+                fs::write(&path, genesis_signature_string)?;
+            }
+
+            writeln!(
+                writer,
+                "Genesis was successfully signed and written to `{}`",
+                path.display()
+            )?;
+        } else if self.binary {
+            writeln!(
+                writer,
+                "SCALE encoded data is not supported for console outputs."
+            )?;
+        } else {
+            writeln!(writer, "{genesis_signature_string}")?;
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use iroha_genesis::{GenesisSignature, GenesisTransaction};
+
+    use super::*;
+    use crate::crypto::AlgorithmArg;
+
+    const GENESIS_JSON_PATH: &str = "../../configs/swarm/genesis.json";
+
+    #[test]
+    fn genesis_signing_works() -> Result<()> {
+        let keypair_config = crypto::Args {
+            algorithm: AlgorithmArg::default(),
+            private_key: None,
+            seed: None,
+            json: true,
+            compact: false,
+        };
+
+        let mut keypair_json = BufWriter::new(Vec::new());
+        keypair_config.run(&mut keypair_json)?;
+        let keypair: KeyPair = serde_json::from_slice(keypair_json.buffer())?;
+
+        let tmp_keypair_json_file = tempfile::NamedTempFile::new()?;
+        fs::write(tmp_keypair_json_file.path(), keypair_json.buffer())?;
+
+        let chain_id_slice = "0123456";
+        let chain_id = ChainId::from(chain_id_slice);
+        let crypto_genesis_config = SignArgs::try_parse_from([
+            "sign",
+            "--chain-id",
+            chain_id_slice,
```

Comment:
```suggestion
```

this should be in the genesis file, not given over CLI

Reply:
Correct me if I wrong, but I don't see genesis file containing chain id.
And it seems like every time we create genesis, be it inside the swarm or core libs, we supply chain id explicitly.
So where do we find it besides taking it over CLI?

Reply:
> Correct me if I wrong, but I don't see genesis file containing chain id.

Yes, add it to the file

Reply:
Umm, duplicating the chain id in the genesis is unfortunate =(

Then we will have to decide what is the source of truth - Chain ID in the config file or in the genesis file? We will also have to validate their equality during config parsing.

So, maybe `--chain-id` is fine? Another option might be to provide a relevant `--config-file`, from where `kagami` can read `chain_id`.